homeassistant/components/frank_energie/coordinator.py

```diff
@@ -0,0 +1,136 @@
+        try:
+            prices_today = await self.__fetch_prices_with_fallback(today, tomorrow)
+            prices_tomorrow = await self.__fetch_prices_with_fallback(
+                tomorrow, day_after_tomorrow
+            )
+
+            data_month_summary = (
+                await self.api.month_summary() if self.api.is_authenticated else None
+            )
+            data_invoices = (
+                await self.api.invoices() if self.api.is_authenticated else None
+            )
+        except RequestException as ex:
+            if str(ex).startswith("user-error:"):
+                raise ConfigEntryAuthFailed from ex
+
+            raise UpdateFailed(ex) from ex
+
+        except AuthException as ex:
+            LOGGER.debug("Authentication tokens expired, trying to renew them (%s)", ex)
+            await self.__try_renew_token()
+            # Tell we have no data, so update coordinator tries again with renewed tokens
+            raise UpdateFailed(ex) from ex
+
+        return DeviceResponseEntry(
+            electricity=prices_today.electricity + prices_tomorrow.electricity,
+            gas=prices_today.gas + prices_tomorrow.gas,
+            month_summary=data_month_summary,
+            invoices=data_invoices,
+        )
+
+    async def __fetch_prices_with_fallback(
+        self, start_date: date, end_date: date
+    ) -> MarketPrices:
+        if not self.api.is_authenticated:
+            return await self.api.prices(start_date, end_date)
+
+        user_prices = await self.api.user_prices(start_date)
+
+        if len(user_prices.gas.all) > 0 and len(user_prices.electricity.all) > 0:
+            # If user_prices are available for both gas and electricity return them
+            return user_prices
+
+        public_prices = await self.api.prices(start_date, end_date)
+
+        # Use public prices if no user prices are available
+        if len(user_prices.gas.all) == 0:
+            LOGGER.info("No gas prices found for user, falling back to public prices")
+            user_prices.gas = public_prices.gas
+
+        if len(user_prices.electricity.all) == 0:
+            LOGGER.info(
+                "No electricity prices found for user, falling back to public prices"
+            )
+            user_prices.electricity = public_prices.electricity
+
+        return user_prices
+
+    async def __try_renew_token(self):
+        try:
+            updated_tokens = await self.api.renew_token()
+
+            data = {
+                CONF_AUTH_TOKEN: updated_tokens.authToken,
+                CONF_REFRESH_TOKEN: updated_tokens.refreshToken,
+            }
+            self.hass.config_entries.async_update_entry(self.entry, data=data)
+
+            LOGGER.debug("Successfully renewed token")
+
+        except AuthException as ex:
+            LOGGER.error("Failed to renew token: %s. Starting user reauth flow", ex)
+            raise ConfigEntryAuthFailed from ex
```

Comment:
You don't have a reauth flow yet, so this would raise an error. I think you should raise `ConfigEntryError`